File: RefactoringTools/RefactoringTools/SyntaxNodeExtensions.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Formatting;
using Microsoft.CodeAnalysis.MSBuild;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace RefactoringTools
{
    public static class SyntaxNodeExtensions
    {
        private static readonly Lazy<MSBuildWorkspace> LazyDefaultWorkspace = new Lazy<MSBuildWorkspace>(() => MSBuildWorkspace.Create());

        public static TSyntaxNode TryFindParentWithinStatement<TSyntaxNode>(this SyntaxNode node, SyntaxKind kind, Func<TSyntaxNode, bool> predicate = null)
            where TSyntaxNode : SyntaxNode
        {
            do
            {
                node = node.Parent;

                if (node.IsStatement() || node.IsMethodOrClassOrNamespace())
                    return null;

                if (node.IsKind(kind))
                {
                    var temp = (TSyntaxNode)node;
                    if (predicate != null)
                    {
                        if (predicate(temp))
                        {
                            return temp;
                        }
                    }
                    else
                    {
                        return temp;
                    }
                }

            } while (true);            
        }

        public static TSyntaxNode TryFindOuterMostParentWithinStatement<TSyntaxNode>(this SyntaxNode node, SyntaxKind kind, Func<TSyntaxNode, bool> predicate = null)
            where TSyntaxNode : SyntaxNode
        {
            TSyntaxNode lastMatched = null;

            do
            {
                node = node.Parent;

                if (node.IsStatement() || node.IsMethodOrClassOrNamespace())
                    return lastMatched;

                if (node.IsKind(kind))
                {
                    var temp = (TSyntaxNode)node;

                    if (predicate != null)
                    {
                        if (predicate(temp))
                        {
                            lastMatched = temp;
                        }
                    }
                    else
                    {
                        lastMatched = temp;
                    }
                }

            } while (true);                                    
        }

        public static BlockSyntax TryFindParentBlock(this SyntaxNode node)
        {
            do
            {
                if (node.IsMethodOrClassOrNamespace())
                    return null;

                node = node.Parent;

            } while (!node.IsKind(SyntaxKind.Block));

            return (BlockSyntax)node;
        }

        public static StatementSyntax TryFindParentStatement(this SyntaxNode node)
        {
            do
            {
                if (node.IsMethodOrClassOrNamespace())
                    return null;

                node = node.Parent;

            } while (!node.IsStatement());

            return (StatementSyntax)node;
        }

        private static bool IsStatement(this SyntaxNode node)
        {
            switch (node.CSharpKind())
            {
                case SyntaxKind.LocalDeclarationStatement:
                case SyntaxKind.ExpressionStatement:
                case SyntaxKind.LabeledStatement:
                case SyntaxKind.GotoStatement:
                case SyntaxKind.GotoCaseStatement:
                case SyntaxKind.GotoDefaultStatement:
                case SyntaxKind.ReturnStatement:
                case SyntaxKind.YieldReturnStatement:
                case SyntaxKind.ThrowStatement:
                case SyntaxKind.WhileStatement:
                case SyntaxKind.DoStatement:
                case SyntaxKind.ForStatement:
                case SyntaxKind.ForEachStatement:
                case SyntaxKind.UsingStatement:
                case SyntaxKind.FixedStatement:
                case SyntaxKind.CheckedStatement:
                case SyntaxKind.UncheckedStatement:
                case SyntaxKind.UnsafeStatement:
                case SyntaxKind.LockStatement:
                case SyntaxKind.IfStatement:
                case SyntaxKind.SwitchStatement:
                case SyntaxKind.TryStatement:
                    return true;
                default:
                    return false;
            }
        }

        private static bool IsMethodOrClassOrNamespace(this SyntaxNode node)
        {
            return node.IsKind(SyntaxKind.MethodDeclaration) || node.IsKind(SyntaxKind.ClassDeclaration) || node.IsKind(SyntaxKind.NamespaceDeclaration);
        }

        public static T Format<T>(this T node) where T : SyntaxNode
        {
            return node.WithAdditionalAnnotations(Formatter.Annotation);
        }
    }
}

File: RefactoringTools/RefactoringTools.VB/SyntaxNodeExtensions.vb
Imports System.Runtime.CompilerServices
Imports Microsoft.CodeAnalysis.Formatting

Module SyntaxNodeExtensions

    <Extension()>
    Private Function IsMethodOrClassOrNamespace(node As SyntaxNode) As Boolean

        Return node.IsKind(SyntaxKind.FunctionBlock) OrElse
               node.IsKind(SyntaxKind.SubBlock) OrElse
               node.IsKind(SyntaxKind.ClassBlock) OrElse
               node.IsKind(SyntaxKind.ModuleBlock) OrElse
               node.IsKind(SyntaxKind.CompilationUnit)
    End Function

    <Extension()>
    Public Function Format(Of T As SyntaxNode)(node As T) As T

        Return node.WithAdditionalAnnotations(Formatter.Annotation)

    End Function

    <Extension()>
    Public Function IsWithin(node As SyntaxNode, span As TextSpan) As Boolean

        Return node.SpanStart >= span.Start AndAlso node.SpanStart <= span.End

    End Function

    <Extension()>
    Public Function TryFindParentWithinStatement(Of TSyntaxNode As SyntaxNode)(
        node As SyntaxNode, kind As SyntaxKind) As TSyntaxNode

        Dim t = New Tuple(Of Integer, String)(42, New String("a"))

        Do
            node = node.Parent

            If (node.IsStatement() OrElse node.IsMethodOrClassOrNamespace()) Then
                Return Nothing
            End If


            If (node.IsKind(kind)) Then

                Dim temp = CType(node, TSyntaxNode)
                Return temp

            End If

        Loop While True

        Return Nothing

    End Function


    '<Extension()>
    'Public Function TryFindParentBlock(node As SyntaxNode) As BlockSyntax
    '    Do
    '        If (node.IsMethodOrClassOrNamespace) Then
    '            Return Nothing
    '        End If

    '        node = node.Parent

    '    Loop While (Not node.IsStatement)

    '    Return node
    'End Function

    <Extension()>
    Public Function TryFindParentStatement(node As SyntaxNode) As StatementSyntax
        Do
            If (node.IsMethodOrClassOrNamespace) Then
                Return Nothing
            End If

            node = node.Parent

        Loop While (Not node.IsStatement)

        Return node
    End Function

    <Extension()>
    Public Function IsStatement(node As SyntaxNode) As Boolean

        Select Case node.VisualBasicKind
            Case SyntaxKind.EmptyStatement,
                 SyntaxKind.EndIfStatement,
                 SyntaxKind.EndUsingStatement,
                 SyntaxKind.EndWithStatement,
                 SyntaxKind.EndSelectStatement,
                 SyntaxKind.EndStructureStatement,
                 SyntaxKind.EndEnumStatement,
                 SyntaxKind.EndInterfaceStatement,
                 SyntaxKind.EndClassStatement,
                 SyntaxKind.EndModuleStatement,
                 SyntaxKind.EndNamespaceStatement,
                 SyntaxKind.EndSubStatement,
                 SyntaxKind.EndFunctionStatement,
                 SyntaxKind.EndGetStatement,
                 SyntaxKind.EndSetStatement,
                 SyntaxKind.EndPropertyStatement,
                 SyntaxKind.EndOperatorStatement,
                 SyntaxKind.EndEventStatement,
                 SyntaxKind.EndAddHandlerStatement,
                 SyntaxKind.EndRemoveHandlerStatement,
                 SyntaxKind.EndRaiseEventStatement,
                 SyntaxKind.EndWhileStatement,
                 SyntaxKind.EndTryStatement,
                 SyntaxKind.EndSyncLockStatement,
                 SyntaxKind.OptionStatement,
                 SyntaxKind.ImportsStatement,
                 SyntaxKind.NamespaceStatement,
                 SyntaxKind.InheritsStatement,
                 SyntaxKind.ImplementsStatement,
                 SyntaxKind.ModuleStatement,
                 SyntaxKind.StructureStatement,
                 SyntaxKind.InterfaceStatement,
                 SyntaxKind.ClassStatement,
                 SyntaxKind.EnumStatement,
                 SyntaxKind.SubStatement,
                 SyntaxKind.FunctionStatement,
                 SyntaxKind.SubNewStatement,
                 SyntaxKind.DeclareSubStatement,
                 SyntaxKind.DeclareFunctionStatement,
                 SyntaxKind.DelegateSubStatement,
                 SyntaxKind.DelegateFunctionStatement,
                 SyntaxKind.EventStatement,
                 SyntaxKind.OperatorStatement,
                 SyntaxKind.PropertyStatement,
                 SyntaxKind.GetAccessorStatement,
                 SyntaxKind.SetAccessorStatement,
                 SyntaxKind.AddHandlerAccessorStatement,
                 SyntaxKind.RemoveHandlerAccessorStatement,
                 SyntaxKind.RaiseEventAccessorStatement,
                 SyntaxKind.AttributesStatement,
                 SyntaxKind.ExpressionStatement,
                 SyntaxKind.PrintStatement,
                 SyntaxKind.LocalDeclarationStatement,
                 SyntaxKind.LabelStatement,
                 SyntaxKind.GoToStatement,
                 SyntaxKind.StopStatement,
                 SyntaxKind.EndStatement,
                 SyntaxKind.ExitDoStatement,
                 SyntaxKind.ExitForStatement,
                 SyntaxKind.ExitSubStatement,
                 SyntaxKind.ExitFunctionStatement,
                 SyntaxKind.ExitOperatorStatement,
                 SyntaxKind.ExitPropertyStatement,
                 SyntaxKind.ExitTryStatement,
                 SyntaxKind.ExitSelectStatement,
                 SyntaxKind.ExitWhileStatement,
                 SyntaxKind.ContinueWhileStatement,
                 SyntaxKind.ContinueDoStatement,
                 SyntaxKind.ContinueForStatement,
                 SyntaxKind.ReturnStatement,
                 SyntaxKind.IfStatement,
                 SyntaxKind.ElseIfStatement,
                 SyntaxKind.ElseStatement,
                 SyntaxKind.TryStatement,
                 SyntaxKind.CatchStatement,
                 SyntaxKind.FinallyStatement,
                 SyntaxKind.OnErrorGoToZeroStatement,
                 SyntaxKind.OnErrorGoToMinusOneStatement,
                 SyntaxKind.OnErrorGoToLabelStatement,
                 SyntaxKind.OnErrorResumeNextStatement,
                 SyntaxKind.ResumeStatement,
                 SyntaxKind.ResumeLabelStatement,
                 SyntaxKind.ResumeNextStatement,
                 SyntaxKind.SelectStatement,
                 SyntaxKind.CaseStatement,
                 SyntaxKind.CaseElseStatement,
                 SyntaxKind.SyncLockStatement,
                 SyntaxKind.DoStatement,
                 SyntaxKind.LoopStatement,
                 SyntaxKind.WhileStatement,
                 SyntaxKind.ForStatement,
                 SyntaxKind.ForEachStatement,
                 SyntaxKind.NextStatement,
                 SyntaxKind.UsingStatement,
                 SyntaxKind.ThrowStatement,
                 SyntaxKind.SimpleAssignmentStatement,
                 SyntaxKind.MidAssignmentStatement,
                 SyntaxKind.AddAssignmentStatement,
                 SyntaxKind.SubtractAssignmentStatement,
                 SyntaxKind.MultiplyAssignmentStatement,
                 SyntaxKind.DivideAssignmentStatement,
                 SyntaxKind.IntegerDivideAssignmentStatement,
                 SyntaxKind.ExponentiateAssignmentStatement,
                 SyntaxKind.LeftShiftAssignmentStatement,
                 SyntaxKind.RightShiftAssignmentStatement,
                 SyntaxKind.ConcatenateAssignmentStatement,
                 SyntaxKind.CallStatement,
                 SyntaxKind.AddHandlerStatement,
                 SyntaxKind.RemoveHandlerStatement,
                 SyntaxKind.RaiseEventStatement,
                 SyntaxKind.WithStatement,
                 SyntaxKind.ReDimStatement,
                 SyntaxKind.ReDimPreserveStatement
                Return True
            Case Else
                Return False

        End Select



    End Function

End Module
